File: apps/web/lib/api/chapters.ts
/**
 * Typed endpoint functions for the Chapters resource.
 *
 * Chapters are nested under a book (`apps/api/app/api/v1/chapters.py`, prefix
 * `/books/{book_id}/chapters` under `/api/v1`):
 *   GET    /books/{bid}/chapters          → ChapterRead[]
 *   POST   /books/{bid}/chapters          → ChapterRead (201)
 *   GET    /books/{bid}/chapters/{cid}    → ChapterRead
 *   PATCH  /books/{bid}/chapters/{cid}    → ChapterRead
 *   DELETE /books/{bid}/chapters/{cid}    → 204
 *   POST   /books/{bid}/chapters/reorder  → ChapterRead[] (body: { order })
 *
 * Responses are validated with Zod before reaching the UI, so a contract drift
 * surfaces as a thrown error rather than a silent shape mismatch.
 */
import { apiFetch } from "./client";
import {
  chapterCreateSchema,
  chapterListSchema,
  chapterReadSchema,
  chapterReorderSchema,
  chapterUpdateSchema,
  type ChapterCreate,
  type ChapterRead,
  type ChapterReorder,
  type ChapterUpdate,
} from "./types";

/** List the chapters belonging to a book (backend orders by `order_index`). */
export async function listChapters(bookId: string): Promise<ChapterRead[]> {
  const data = await apiFetch<unknown>(`/books/${bookId}/chapters`);
  return chapterListSchema.parse(data);
}

/** Fetch a single chapter within a book. */
export async function getChapter(
  bookId: string,
  chapterId: string,
): Promise<ChapterRead> {
  const data = await apiFetch<unknown>(
    `/books/${bookId}/chapters/${chapterId}`,
  );
  return chapterReadSchema.parse(data);
}

/** Create a chapter under a book. The payload is validated against the contract. */
export async function createChapter(
  bookId: string,
  input: ChapterCreate,
): Promise<ChapterRead> {
  const body = chapterCreateSchema.parse(input);
  const data = await apiFetch<unknown>(`/books/${bookId}/chapters`, {
    method: "POST",
    body,
  });
  return chapterReadSchema.parse(data);
}

/** Patch a chapter (title / summary / status / order_index). */
export async function updateChapter(
  bookId: string,
  chapterId: string,
  input: ChapterUpdate,
): Promise<ChapterRead> {
  const body = chapterUpdateSchema.parse(input);
  const data = await apiFetch<unknown>(
    `/books/${bookId}/chapters/${chapterId}`,
    { method: "PATCH", body },
  );
  return chapterReadSchema.parse(data);
}

/** Delete a chapter (cascades to its scenes server-side). */
export async function deleteChapter(
  bookId: string,
  chapterId: string,
): Promise<void> {
  await apiFetch<unknown>(`/books/${bookId}/chapters/${chapterId}`, {
    method: "DELETE",
  });
}

/**
 * Reorder a book's chapters. `order` is the full list of chapter ids in their
 * new sequence; the backend assigns `order_index` by position and returns the
 * re-sorted chapter list.
 */
export async function reorderChapters(
  bookId: string,
  order: string[],
): Promise<ChapterRead[]> {
  const body: ChapterReorder = chapterReorderSchema.parse({ order });
  const data = await apiFetch<unknown>(`/books/${bookId}/chapters/reorder`, {
    method: "POST",
    body,
  });
  return chapterListSchema.parse(data);
}

File: apps/api/app/api/v1/chapters.py
import uuid
from fastapi import APIRouter, Depends, HTTPException, status
from sqlalchemy import select
from sqlalchemy.ext.asyncio import AsyncSession

from app.core.deps import get_current_user, get_db
from app.models.book import Book
from app.schemas.chapter import ChapterCreate, ChapterRead, ChapterReorder, ChapterUpdate
from app.services.crud_chapter import (
    create_chapter,
    delete_chapter,
    get_chapter,
    list_chapters,
    reorder_chapters,
    update_chapter,
)

router = APIRouter(prefix="/books/{book_id}/chapters", tags=["chapters"])


async def _get_book_or_404(book_id: uuid.UUID, db: AsyncSession):
    result = await db.execute(select(Book).where(Book.id == book_id))
    book = result.scalar_one_or_none()
    if book is None:
        raise HTTPException(status_code=status.HTTP_404_NOT_FOUND, detail="Book not found")
    return book


@router.post("", response_model=ChapterRead, status_code=status.HTTP_201_CREATED)
async def create(
    book_id: uuid.UUID,
    data: ChapterCreate,
    db: AsyncSession = Depends(get_db),
    _: str = Depends(get_current_user),
) -> ChapterRead:
    await _get_book_or_404(book_id, db)
    return await create_chapter(db, book_id, data)


@router.get("", response_model=list[ChapterRead])
async def list_all(
    book_id: uuid.UUID,
    db: AsyncSession = Depends(get_db),
    _: str = Depends(get_current_user),
) -> list[ChapterRead]:
    await _get_book_or_404(book_id, db)
    return await list_chapters(db, book_id)


@router.post("/reorder", response_model=list[ChapterRead])
async def reorder(
    book_id: uuid.UUID,
    data: ChapterReorder,
    db: AsyncSession = Depends(get_db),
    _: str = Depends(get_current_user),
) -> list[ChapterRead]:
    await _get_book_or_404(book_id, db)
    try:
        return await reorder_chapters(db, book_id, data.order)
    except ValueError as e:
        raise HTTPException(status_code=status.HTTP_400_BAD_REQUEST, detail=str(e))


@router.get("/{chapter_id}", response_model=ChapterRead)
async def get_one(
    book_id: uuid.UUID,
    chapter_id: uuid.UUID,
    db: AsyncSession = Depends(get_db),
    _: str = Depends(get_current_user),
) -> ChapterRead:
    await _get_book_or_404(book_id, db)
    chapter = await get_chapter(db, book_id, chapter_id)
    if chapter is None:
        raise HTTPException(status_code=status.HTTP_404_NOT_FOUND, detail="Chapter not found")
    return chapter


@router.patch("/{chapter_id}", response_model=ChapterRead)
async def update(
    book_id: uuid.UUID,
    chapter_id: uuid.UUID,
    data: ChapterUpdate,
    db: AsyncSession = Depends(get_db),
    _: str = Depends(get_current_user),
) -> ChapterRead:
    await _get_book_or_404(book_id, db)
    chapter = await get_chapter(db, book_id, chapter_id)
    if chapter is None:
        raise HTTPException(status_code=status.HTTP_404_NOT_FOUND, detail="Chapter not found")
    return await update_chapter(db, chapter, data)


@router.delete("/{chapter_id}", status_code=status.HTTP_204_NO_CONTENT)
async def delete(
    book_id: uuid.UUID,
    chapter_id: uuid.UUID,
    db: AsyncSession = Depends(get_db),
    _: str = Depends(get_current_user),
) -> None:
    await _get_book_or_404(book_id, db)
    chapter = await get_chapter(db, book_id, chapter_id)
    if chapter is None:
        raise HTTPException(status_code=status.HTTP_404_NOT_FOUND, detail="Chapter not found")
    await delete_chapter(db, chapter)
